coderd/insights.go

```diff
@@ -43,3 +51,390 @@ func (api *API) deploymentDAUs(rw http.ResponseWriter, r *http.Request) {
 	}
 	httpapi.Write(ctx, rw, http.StatusOK, resp)
 }
+
+// @Summary Get insights about user latency
+// @ID get-insights-about-user-latency
+// @Security CoderSessionToken
+// @Produce json
+// @Tags Insights
+// @Success 200 {object} codersdk.UserLatencyInsightsResponse
+// @Router /insights/user-latency [get]
+func (api *API) insightsUserLatency(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	if !api.Authorize(r, rbac.ActionRead, rbac.ResourceDeploymentValues) {
+		httpapi.Forbidden(rw)
+		return
+	}
+
+	p := httpapi.NewQueryParamParser().
+		Required("start_time").
+		Required("end_time")
+	vals := r.URL.Query()
+	var (
+		// The QueryParamParser does not preserve timezone, so we need
+		// to parse the time ourselves.
+		startTimeString = p.String(vals, "", "start_time")
+		endTimeString   = p.String(vals, "", "end_time")
+		templateIDs     = p.UUIDs(vals, []uuid.UUID{}, "template_ids")
+	)
+	p.ErrorExcessParams(vals)
+	if len(p.Errors) > 0 {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message:     "Query parameters have invalid values.",
+			Validations: p.Errors,
+		})
+		return
+	}
+
+	startTime, endTime, ok := parseInsightsStartAndEndTime(ctx, rw, startTimeString, endTimeString)
+	if !ok {
+		return
+	}
+
+	rows, err := api.Database.GetUserLatencyInsights(ctx, database.GetUserLatencyInsightsParams{
+		StartTime:   startTime,
+		EndTime:     endTime,
+		TemplateIDs: templateIDs,
+	})
+	if err != nil {
+		httpapi.InternalServerError(rw, err)
+		return
+	}
+
+	// Fetch all users so that we can still include users that have no
```

Comment:
This is something I would reconsider. Let's assume that we're going to integrate with a company account system using OIDC. We're going to fetch ALL accounts, and I suppose that 80% may be not relevant (sales, account managers, etc.)

I'm wondering if we can limit this query somehow. Do we need data from inactive users?

Thinking loud: is it possible to fetch insights first, and then just check users that were playing with Coder at that time?

Reply:
I'm not sure how an inactive user would be able to generate data in the first place. At a surface level, it seems to make sense to exclude them. We could add a parameter `activeUsersOnly` but we may not actually need it.

Reply:
An inactive user can generate data if they became inactive later. We do currently filter out both inactive and deleted users from the result, btw.

One option is to only include the users did use their workspace, but I'm worried that someone looking at the data may wonder why a user sometimes is shown and other times not.

Another option may be to filter out users that don't have a workspace (in addition to current filters).

> Thinking loud: is it possible to fetch insights first, and then just check users that were playing with Coder at that time?

This is kind of what we're doing right now, except we're also fetching all users and attaching insights to them where appropriate.

Reply:
I think that the edge case is when user was very active, and then admit deactivated the account.

Reply:
> This is kind of what we're doing right now, except we're also fetching all users and attaching insights to them where appropriate.

> also fetching all users and attaching insights

actually I was thinking if we can filter only relevant users.

Reply:
> actually I was thinking if we can filter only relevant users.

I'm definitely for filtering on only relevant users, but we'd need to lock down a definition for what a relevant user is.

For instance, how about this?
> If a user has a workspace and either used it in the timeframe OR is not an inactive user, include. When filtering on templates, the user must have a workspace in said template.

Reply:
This has been amended so that we only return users that generated latency data for the interval: [`fc2157d` (#8519)](https://github.com/coder/coder/pull/8519/commits/fc2157d7ee0e777dd247ea661607becaeb384b9a)